### PR TITLE
PIP-117: Change Pulsar standalone defaults

### PR DESCRIPTION
**Pull request by merlimat**
Master Issue: #13302

### Motivation

 * Change standalone metadata impl from ZooKeeper to RocksDB

 * Pulsar functions package backend -->  Change from using DistributedLog to using local filesystem, storing the
jars directly in the data folder instead of uploading them into BK.

 * Pulsar functions state store implementation --> Change the state store to be backed by a MetadataStore based backed,
with the RocksDB implementation.

 * Table Service --> Do not start BK table service by default

### CODE REVIEW

**Review by Shoothzj — APPROVED** [review]

**Review by Demogorgon314 — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandalone.java [hunk 1]
```diff
@@ -246,7 +259,10 @@ public boolean isHelp() {
     @Parameter(names = { "-h", "--help" }, description = "Show this help message")
     private boolean help = false;
 
+    private boolean usingNew211Defaults;
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandalone.java:262]: Using version smells a little from my point ofbview.
What about using PIP117 as reference?
usingPIP117Defaults
or we can invert the flag and name it 'useLegscyDefaults' and cite this PIP with a link

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandalone.java [hunk 2]
```diff
@@ -246,7 +259,10 @@ public boolean isHelp() {
...
+
     public void start() throws Exception {
+        usingNew211Defaults = !Paths.get(zkDir).toFile().exists();
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandalone.java:265]: Os this is true maybe we could log a line explaining that we found  the directory and we are switching to the old defaults

**Review by eolivelli — CHANGES_REQUESTED** [review]
Some thought (I voted the PIP but now that we are close to committing the patch some doibts camento my mind)
Disabling zookeeper by default may have some nasty impact of integration tests of many applications and that run the same test against many versions of Pulsar.

I wonder if there is some easy way, especially for docker users, like passing a single ENV entry of telling Pulsar standalone to run with zookeeper.
It will be quite difficult to support probably but I am now imagining the pain for making integration tests of many applications work with different versions of Pulsar

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandalone.java (hunk 2 above)
Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandalone.java:265]: A warning log would be great when both `--metadata-dir` is provided and `zkDir` exists.

**Review by Jason918 — APPROVED** [review]
LGTM. Left a minor suggestion.

**Comment by merlimat** [comment]
> Some thought (I voted the PIP but now that we are close to committing the patch some doibts camento my mind) Disabling zookeeper by default may have some nasty impact of integration tests of many applications and that run the same test against many versions of Pulsar.
> 
> I wonder if there is some easy way, especially for docker users, like passing a single ENV entry of telling Pulsar standalone to run with zookeeper. It will be quite difficult to support probably but I am now imagining the pain for making integration tests of many applications work with different versions of Pulsar

I'm not sure I'm following here. The external interface of Pulsar does not change whether its metadata is stored in ZK or RocksDb. 

I agree though that it's probably a good idea to leave a CLI flag to force the ZK mode, to leave the option open, but that wouldn't be the default.

**Review by eolivelli — APPROVED** [review]
Lgtm

Thank you!

**Review by mattisonchao — APPROVED** [review]

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java [hunk 3]
```diff
@@ -81,4 +81,12 @@ public static String removeIdentifierFromMetadataURL(String metadataURL) {
         }
         return metadataURL;
     }
+
+    public static boolean isBasedOnZookeeper(String metadataURL) {
+        if (!metadataURL.contains("://")) {
```
Comment (tisonkun) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java:86]: @merlimat it seems that metadata url doesn't contains `//` part?

From the codebase a metadata url to zk is `zk:127.0.0.1:2181` and a metadata url to memory is `memory:local`.

**Pull request merged**